### ISSUE TITLE
Can we specify that `{}` has type `LinkedHashMap`?

### ISSUE THREAD

**Issue by scheglov** [dart-lang/language#2721]

**Comment by lrhn** [comment]
I assume you mean to specify that the runtime type of a non-constant map literal is `LinkedHashMap`. If so, yes, we should state that. It's true, and people are likely depending on it (even if they shouldn't).
In particular, `LinkedHashMap` is one of the types allowed to be sent to other isolate groups. (But we could also just say that the maps created by map literals can be sent. We need to say that anyway, because constant map literals are not `LinkedHashMap`s).

The static type should still be just `Map`, because being a `LinkedHashMap` is an implementation detail, and you should program to the interface, not the implementation type.

**Comment by eernstg** [comment]
I tend to think that it would be more useful to state some properties, e.g., that the value of a map literal preserves the insertion order during iterations using `keys` or `entries`, or something like that&mdash;whatever is true and useful.

I'm not so convinced that it is actually helpful to cast in concrete that the run-time type of the map is a subtype of `LinkedHashMap`.

How would that help anybody? What will we do if we discover that some other implementation than LinkedHash is better in some situations? Perhaps some map literals will evaluate to instances of `LinkedHashMap`, and others will evaluate to `NewFancyDataStructureMap`; this could be in the same program execution or on different platforms.

I agree that the static type of map literals can't change, that would be too breaking.

**Comment by lrhn** [comment]
I think the ship has sailed on changing the runtime type of non-constant map literals. 
Too many people know and expect them to be `LinkedHashMap`s. Even if their code shouldn't reflect that, ... well, it still does. Sometimes, I think it's used as implicit documentation that a map is insertion-ordered.

(If I designed the platform libraries today, `LinkedHashMap` would not be a public class, it'd just be what `Map` constructors forwarded to. But we didn't do that.)

**Comment by rrousselGit** [comment]
> (If I designed the platform libraries today, LinkedHashMap would not be a public class, it'd just be what Map constructors forwarded to. But we didn't do that.)

There's some value in using the `LinkedHashMap` type instead of `Map` directly.   
If some code expects the order to be preserved, typing the map as "linked" guarantees that the input is sane. Otherwise, a non-linked map could be sent to code that expects a linked map.

So I think `LinkedHashMap` being public is fine. 

No need to be able to do `LinkedHashMap map = {}` though. That would of course be too breaking. In the few cases I had to use `LinkedHashMap` explicitly, I didn't mind doing `LinkedHashMap map = LinkedHashMap()`.

**Comment by Cat-sushi** [comment]
The lint  also should be reconsidered.

This is my Least Recently Used Cache implementation.
I have to comment `ignore: prefer_collection_literals` to suppress info.
```dart
import 'dart:collection';

typedef Key = int;
typedef Value = String;

class LruCache {
  final int size;
  // ignore: prefer_collection_literals
  final _map = LinkedHashMap<Key, Value>();

  LruCache(this.size);

  Value? operator [](Key key) {
    var value = _map.remove(key);
    if (value == null) {
      return null;
    }
    _map[key] = value;
    return value;
  }

  operator []=(Key key, Value value) {
    var rce = _map.remove(key) ?? value;
    _map[key] = rce;
    if (_map.length > size) {
      _map.remove(_map.keys.first);
    }
  }
}
```

**Comment by lrhn** [comment]
The [spec says](https://github.com/dart-lang/language/blob/master/specification/dartLangSpec.tex#L11093):
> A map is ordered: iteration over the keys, values, or key/value pairs
> occurs in the order in which the keys were added to the set.

That's not true in general, only for the maps created by map literals and the `LinkedHashMap` type (which document itself as being insertion-ordered). (Also "to the set" should probably be "to the map".)

Maybe say "The map of a map literal is ordered: ..." and "This also applies to keys added after literal was evaluated."

But if the maps created by map literals are insertion-ordered, then the
```dart
  // ignore: prefer_collection_literals
  final _map = LinkedHashMap<Key, Value>();
```
lint is fine, there is no need to insist on `LinkedHashMap`. A `<Key, Value>{}` would work the same.

**Comment by Cat-sushi** [comment]
@lrhn 
> there is no need to insist on LinkedHashMap. A <Key, Value>{} would work the same.

Thank you for your explanation.

Do you mean also that `LinkedHashMap` doesn't have additional features/ characteristic compared to `Map` from map literal?
In other words, `LinkedHashMap` is purely a implementation detail, so, you said that it should not be a public class?

**Comment by lrhn** [comment]
Yes, the interface API of `LinkedHashMap` is precisely the same as for `Map`. If you look at [the docs](https://api.dart.dev/be/175791/dart-collection/LinkedHashMap-class.html), every instance member is "inherited". 
It has different constructors, but after the value is created, there is no longer any value in knowing or remembering that it's a `LinkedHashMap`. 
It's an insertion-ordered map with expected amortized constant-time operations. So are map literals (we should say the "expected amortized constant-time operations" part too). We have chosen to implement (non-constant) map literals using `LinkedHashMap`s because they satisfy the requirements. It's not a coincidence, and we could also just specify that that's what non-constant map literals do, because it's unlikely that we'll change that.
